### PR TITLE
Refactor of code base

### PR DESCRIPTION
**Pull request by rkokkelk**
So this MR is a large refactor of the entire code base. Whilst developing new PR, I've encountered the following in the initial code base:

* Various different methods for accomplishing the same (e.g. calculating number of CSV lines, getting the header)
* A lot of duplicate code
* No unit tests
* No clear distinction between utility functions, front-end / back-end functionality

These issues made it quite difficult to add new contributions. Because of this I've decided to propose a total rewrite of the back-end functionality. The main focus was:

* Implement Model/View/Controller (MV) logic in the backend
* Reduce all duplicate code
* Add some basic unit tests

This resulted in the following major changes:

* Moved a lot of functionality to CSV / CSVLine classes. These classes hold the majority of the functionality
* Add InvalidCellException class for ease of use of validation / uploading CSV files
* Reduced the `backend.py` functionality to only contain 'controller' functionality. 
* Moved all utility functions to `lib.util`. 
* Removed all functionality which might be handled by Flask (config management, template generation, URL generation,  setting response types)
* Ensure that HTML are generated using Flask templates instead of custom code.

Because the rework is so extensive I've not merged it by default but would like some feedback on the work that has been done before I actually merge it into the master.

### CODE REVIEW

File: intelmq_webinput_csv/app.py [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+
+STATIC_FILES = {
+    'js/preview.js': None,
+    'js/upload.js': None,
+    'preview.html': None,
+    'index.html': None,
+}
+
+for static_file in STATIC_FILES.keys():
+    filename = pkg_resources.resource_filename('intelmq_webinput_csv', 'static/%s' % static_file)
```
Comment (waldbauer-certat) [inline intelmq_webinput_csv/app.py:58]: f-string would be more readable

**Review by waldbauer-certat — CHANGES_REQUESTED** [review]

File: tests/test_models/test_csv.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+    @pytest.fixture
+    def file1(self, tmp_path):
+        data1 = \
+"""column1,column2,column3
+line1.1,line1.2,line1.3
+line2.1,line2.2,line2.3
+line3.1,line3.2,line3.3
+"""
+        file_path = tmp_path / 'input.csv'
```
[the remark this answers is not in the document]
Comment (rkokkelk) [reply tests/test_models/test_csv.py:18]: This is default pathlib functionality for joining paths without relying on underlying OS (Should work for both Windows, Linux)

* https://docs.python.org/3/library/pathlib.html

File: tests/base.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    def get_harmonization_config(self, file_name: str='harmonization.conf') -> dict:
+        """ Load harmonization config
+
+        Parameters:
+            file_name: name of harmonization config to use
+        
+        Returns:
+            Dict of parsed harmonization config
+        """
+        configuration_filepath = self.get_fixtures_path() / file_name
```
[the remark this answers is not in the document]
Comment (rkokkelk) [reply tests/base.py:34]: This is default pathlib functionality for joining paths without relying on underlying OS (Should work for both Windows, Linux)

* https://docs.python.org/3/library/pathlib.html

File: intelmq_webinput_csv/lib/util.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        response.mimetype = 'application/json'
+        response.headers['Content-Type'] = "text/json; charset=utf-8"
+    if content_type:
+        response.headers['Content-Type'] = content_type
+    response.headers['Access-Control-Allow-Origin'] = "*"
+    return response
+
+
+def get_temp_file(filename: str = 'webinput_csv.csv') -> Path:
+    return Path(VAR_STATE_PATH) / filename
```
[the remark this answers is not in the document]
Comment (rkokkelk) [reply intelmq_webinput_csv/lib/util.py:127]: This is default pathlib functionality for joining paths without relying on underlying OS (Should work for both Windows, Linux)

* https://docs.python.org/3/library/pathlib.html

File: intelmq_webinput_csv/app.py (hunk 1 above)
Comment (rkokkelk) [reply intelmq_webinput_csv/app.py:58]: This is still leftover from the initial code base but will fix.

File: intelmq_webinput_csv/app.py [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+def form():
+    response = make_response(STATIC_FILES['index.html'])
+    response.mimetype = 'text/html'
+    response.headers['Content-Type'] = "text/html; charset=utf-8"
+    return response
+
+
+@app.route('/plugins/<path:page>')
+def plugins(page):
+    filename = pkg_resources.resource_filename('intelmq_webinput_csv', 'static/plugins/%s' % page)
```
[the remark this answers is not in the document]
Comment (rkokkelk) [reply intelmq_webinput_csv/app.py:90]: This is still leftover from the initial code base but will fix.

File: intelmq_webinput_csv/templates/preview.html [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+                    </div>
+                    <div class="control">
+                        <button id="submitButton" class="button is-success" type="button" v-on:click="submitButtonClicked()">Submit</button>
+                    </div>
+                </div>
+            </div>
+        </section>
+    </form>
+</div>
+{% endblock %}
```
Comment (waldbauer-certat) [inline intelmq_webinput_csv/templates/preview.html:132]: missing newline

**Review by waldbauer-certat — CHANGES_REQUESTED** [review]
LGTM, new lines are still missing

File: intelmq_webinput_csv/templates/preview.html (hunk 6 above)
Comment (rkokkelk) [reply intelmq_webinput_csv/templates/preview.html]: Done

**Comment by wagner-intevation** [comment]
We very much welcome the rewrite of the backend. The original code was a proof of concept that worked "too well" to be replaced.
OTOH, I don't particularly appreciate that we have diverging code bases. Can we get efforts combined?
https://github.com/certat/intelmq-webinput-csv/pull/80#issuecomment-1293793841

**Pull request merged**